In [23]:
# DSC 540
# Week 5 & 6
# Summer 2024
# Term Project Milestone 3
# Created By Matthew Heinen
# 7/6/24

# Weather and Carlos Santana’s Effectiveness
# DSC 540: Milestone 3 

## Read in The Table from Wikipedia. 

In [2]:
# Import packages. 

import pandas as pd

In [4]:
# Use read_html to read in the list of tables in wiki page. 
wiki_list_dfs = pd.read_html("https://en.wikipedia.org/wiki/List_of_current_Major_League_Baseball_stadiums",
                    header=0)

In [5]:
# Check length of the list that is returned
# This is the number of tables in the webpage
len(wiki_list_dfs)

13

In [8]:
# Loop through each table in the wiki df.
# print the shape of each data frame.
# this will help us identify the proper table for out dataset.
for t in wiki_list_dfs:
    print("Table Shape: ", t.shape)

Table Shape:  (30, 10)
Table Shape:  (1, 7)
Table Shape:  (3, 7)
Table Shape:  (19, 2)
Table Shape:  (2, 2)
Table Shape:  (2, 2)
Table Shape:  (8, 2)
Table Shape:  (0, 2)
Table Shape:  (8, 2)
Table Shape:  (2, 2)
Table Shape:  (2, 2)
Table Shape:  (5, 2)
Table Shape:  (2, 2)


It looks like the first table element in the list is the table that we need there are 30 rows, one for every Major League team. 

In [11]:
# Select the first element, index 0
# Assign this table to new df and display. 
wiki = wiki_list_dfs[0]

# Visualize dataframe. 
wiki

Image                             Name    Capacity  \
0     NaN           American Family Field‡   41,900[1]   
1     NaN                    Angel Stadium   45,517[2]   
2     NaN                    Busch Stadium   45,494[3]   
3     NaN                     Chase Field‡   48,405[4]   
4     NaN                       Citi Field   41,922[5]   
5     NaN               Citizens Bank Park   42,792[6]   
6     NaN                    Comerica Park   41,083[7]   
7     NaN                      Coors Field   50,144[8]   
8     NaN                   Dodger Stadium   56,000[9]   
9     NaN                      Fenway Park  37,755[10]   
10    NaN                Globe Life Field‡  40,300[11]   
11    NaN         Great American Ball Park       43500   
12    NaN            Guaranteed Rate Field       40615   
13    NaN                 Kauffman Stadium  37,903[12]   
14    NaN                  LoanDepot Park‡       36742   
15    NaN                Minute Maid Park‡  41,168[14]   
16    NaN                   Nationals Park  41,373[16]   
17    NaN  Oakland–Alameda County Coliseum  46,847[17]   
18    NaN                      Oracle Park  41,265[18]   
19    NaN      Oriole Park at Camden Yards       44970   
20    NaN                       Petco Park  39,860[19]   
21    NaN                         PNC Park  38,747[20]   
22    NaN                Progressive Field  34,830[21]   
23    NaN                   Rogers Centre‡  39,150[22]   
24    NaN                   T-Mobile Park‡  47,929[23]   
25    NaN                     Target Field  38,544[24]   
26    NaN                 Tropicana Field†  25,000[25]   
27    NaN                      Truist Park  41,084[26]   
28    NaN                    Wrigley Field  41,649[27]   
29    NaN                   Yankee Stadium  46,537[28]   

                      Location          Surface                       Team  \
0         Milwaukee, Wisconsin            Grass          Milwaukee Brewers   
1          Anaheim, California            Grass         Los Angeles Angels   
2          St. Louis, Missouri            Grass        St. Louis Cardinals   
3             Phoenix, Arizona  Artificial turf       Arizona Diamondbacks   
4             Queens, New York            Grass              New York Mets   
5   Philadelphia, Pennsylvania            Grass      Philadelphia Phillies   
6            Detroit, Michigan            Grass             Detroit Tigers   
7             Denver, Colorado            Grass           Colorado Rockies   
8      Los Angeles, California            Grass  Los Angeles Dodgers[nb 2]   
9        Boston, Massachusetts            Grass       Boston Red Sox[nb 3]   
10            Arlington, Texas  Artificial turf              Texas Rangers   
11            Cincinnati, Ohio            Grass            Cincinnati Reds   
12           Chicago, Illinois            Grass          Chicago White Sox   
13       Kansas City, Missouri            Grass         Kansas City Royals   
14              Miami, Florida  Artificial turf              Miami Marlins   
15              Houston, Texas            Grass             Houston Astros   
16            Washington, D.C.            Grass       Washington Nationals   
17         Oakland, California            Grass          Oakland Athletics   
18   San Francisco, California            Grass       San Francisco Giants   
19         Baltimore, Maryland            Grass          Baltimore Orioles   
20       San Diego, California            Grass           San Diego Padres   
21    Pittsburgh, Pennsylvania            Grass         Pittsburgh Pirates   
22             Cleveland, Ohio            Grass        Cleveland Guardians   
23            Toronto, Ontario  Artificial turf          Toronto Blue Jays   
24         Seattle, Washington            Grass           Seattle Mariners   
25      Minneapolis, Minnesota            Grass            Minnesota Twins   
26     St. Petersburg, Florida  Artificial turf             Tampa Bay Rays   
27         Cumberland

We have sucessfully read the wikipedia page into the wiki dataframe with all necessary information. 

## Read in The Table from Steamheads.

This table is essential in adding latitude, longitude and altitude. Latitude and longitude are essential in calling weather data from the API. 

In [34]:
# Import packages
import pandas as pd 

In [35]:
# Use read_html to read in list of tables from steamheads page. 
steam_list_dfs = pd.read_html("https://www.seamheads.com/ballparks/index.php",
                    header=0)

In [36]:
# Check length of the list that is returned
# This is the number of tables in the webpage
len(steam_list_dfs)

1

In [37]:
# Verify the shape of the one table in the steam_list_dfs list. 
# Use index 0 
steam_list_dfs[0].shape

(261, 11)

As we can see here there are 261 rows in in the df. This is correct because this table contains ever stadium in the United States where a professional baseball game has been played. Not just Major League stadiums. 

In [39]:
# Select the first element, index 0
# Assign this table to new df and display. 
steam = steam_list_dfs[0]

# Visualize dataframe. 
steam.head(30)

Ballpark           Location    First Gm     Last Gm  \
0                     Fenway Park         Boston, MA  04/20/1912         NaN   
1                   Wrigley Field        Chicago, IL  04/23/1914         NaN   
2            Sportsman's Park III      St. Louis, MO  04/14/1909  05/08/1966   
3                   Tiger Stadium        Detroit, MI  04/20/1912  09/27/1999   
4                Yankee Stadium I       New York, NY  04/18/1923  09/21/2008   
5                 Comiskey Park I        Chicago, IL  07/01/1910  09/30/1990   
6                      Shibe Park   Philadelphia, PA  04/12/1909  10/01/1970   
7                  Dodger Stadium    Los Angeles, CA  04/10/1962         NaN   
8                    Forbes Field     Pittsburgh, PA  06/30/1909  06/28/1970   
9        Angel Stadium of Anaheim        Anaheim, CA  04/19/1966         NaN   
10                  Crosley Field     Cincinnati, OH  04/11/1912  06/24/1970   
11                 Polo Grounds V       New York, NY  06/28/1911  09/18/1963   
12  Oakland-Alameda County Col...        Oakland, CA  04/17/1968         NaN   
13              Cleveland Stadium      Cleveland, OH  07/31/1932  10/03/1993   
14               Kauffman Stadium    Kansas City, MO  04/10/1973         NaN   
15               Griffith Stadium     Washington, DC  04/12/1911  09/21/1961   
16                   Shea Stadium       New York, NY  04/17/1964  09/28/2008   
17                 County Stadium      Milwaukee, WI  04/14/1953  09/28/2000   
18                   Ebbets Field       Brooklyn, NY  04/09/1913  09/24/1957   
19                     Baker Bowl   Philadelphia, PA  05/02/1895  06/30/1938   
20               Busch Stadium II      St. Louis, MO  05/12/1966  10/02/2005   
21               Candlestick Park  San Francisco, CA  04/12/1960  09/30/1999   
22               Memorial Stadium      Baltimore, MD  04/15/1954  10/06/1991   
23                   Braves Field         Boston, MA  08/18/1915  09/21/1952   
24                      Astrodome        Houston, TX  04/12/1965  10/03/1999   
25            Jack Murphy Stadium      San Diego, CA  04/08/1969  09/28/2003   
26                  Rogers Centre       Toronto, ONT  06/05/1989         NaN   
27               Veterans Stadium   Philadelphia, PA  04/10/1971  09/28/2003   
28            Guarantee Rate Park        Chicago, IL  04/18/1991         NaN   
29             Riverfront Stadium     Cincinnati, OH  06/30/1970  09/22/2002   

    Seasons   Games    Lat    Long  Alt Current Historic  
0       112  8771.0  42.35  -71.10   16     Map      Map  
1       110  8634.0  41.95  -87.66  599     Map      Map  
2        58  7022.0  38.66  -90.22  492     Map      Map  
3        88  6879.0  42.33  -83.07  601     Map      Map  
4        84  6581.0  40.83  -73.93   18     Map      Map  
5        81  6247.0  41.83  -87.63  594     Map      Map  
6        62  6055.0  40.00  -75.16  116     Map      Map  
7        62  5230.0  34.07 -118.24  501     Map      Map  
8        62  4760.0  40.44  -79.95  893     Map      Map  
9        58  4595.0  33.80 -117.88  154     Map      Map  
10       59  4544.0  39.12  -84.54  488     Map      Map  
11       49  4490.0  40.83  -73.94   13     Map      Map  
12       56  4414.0  37.75 -122.20    0     Map      Map  
13       60  4196.0  41.51  -81.70  593     Map      Map  
14       51  4009.0  39.05  -94.48  886     Map      Map  
15       51  3945.0  38.92  -77.02  100     Map      Map  
16       45  3737.0  40.76  -73.85   13     Map      Map  
17       46  3484.0  43.03  -87.97  606     Map      Map  
18       45  3447.0  40.66  -73.96   84     Map      Map  
19       44  3218.0  39.99  -75.16   98     Map      Map  
20       40  3174.0  38.62  -90.19  460     Map      Map  
21       40  3173.0  37.71 -122.39  135     Map      Map  
22       38  3006.0  39.33  -76.60  231     Map      Map  
23       38  2871.0  42.35  -71.12   26     Map      Map  
24       35  2774.0  29.68  -95.41   52     Map      Map  
25       3

We have successfully read the steamheads page into the steam dataframe with all necessary information.

# Transformation 1: Remove dagger mark from "Name" column in wiki DataFrame. 

This will ensure that we can join the two DataFrames on the "Name" and "Ballpark" columns respectively. Ensuring that we can join the two DataFrames into one. 

In [42]:
# Use the str.replace method to remove the dagger character from every value in the "Name" column. 
# Use wiki dataset
wiki['Name'] = wiki['Name'].str.replace('‡', '')

In [60]:
# Verify that the dagger is removed by printing every value in the name column. 

# Loop through every value in the Name column. 
for value in wiki['Name']:
    # Print value. 
    print(value)

# Verify the shape of wiki
print(wiki.shape)

American Family Field
Angel Stadium
Busch Stadium
Chase Field
Citi Field
Citizens Bank Park
Comerica Park
Coors Field
Dodger Stadium
Fenway Park
Globe Life Field
Great American Ball Park
Guaranteed Rate Field
Kauffman Stadium
LoanDepot Park
Minute Maid Park
Nationals Park
Oakland–Alameda County Coliseum
Oracle Park
Oriole Park at Camden Yards
Petco Park
PNC Park
Progressive Field
Rogers Centre
T-Mobile Park
Target Field
Tropicana Field†
Truist Park
Wrigley Field
Yankee Stadium
(30, 10)


We have sucessfully removed the dagger charachter from every name entry in the DataFrame. 

# Transformation 2: Eliminate Non-Active ballparks in Steam Dataset. 

This will ensure that we are merging our data down the line properly and referring to the proper ballpark. For example, Yankee Stadium has multiple listings in the steam dataset as there have been multiple Yankee Stadiums historically. 

In [77]:
# Use the isnull() method to filter the steam dataset for ballparks where the last game is null.
# This will return only active ballparks. 
steam = steam[steam['Last Gm'].isnull()]

In [83]:
# Display the resulting DataFrame. 
steam

Ballpark            Location    First Gm Last Gm  \
0                      Fenway Park          Boston, MA  04/20/1912     NaN   
1                    Wrigley Field         Chicago, IL  04/23/1914     NaN   
7                   Dodger Stadium     Los Angeles, CA  04/10/1962     NaN   
9         Angel Stadium of Anaheim         Anaheim, CA  04/19/1966     NaN   
12   Oakland-Alameda County Col...         Oakland, CA  04/17/1968     NaN   
14                Kauffman Stadium     Kansas City, MO  04/10/1973     NaN   
26                   Rogers Centre        Toronto, ONT  06/05/1989     NaN   
28             Guarantee Rate Park         Chicago, IL  04/18/1991     NaN   
30   Oriole Park at Camden Yard...       Baltimore, MD  04/06/1992     NaN   
34               Progressive Field       Cleveland, OH  04/04/1994     NaN   
35                     Coors Field          Denver, CO  04/26/1995     NaN   
39                     Chase Field         Phoenix, AZ  03/31/1998     NaN   
40                 Tropicana Field  St. Petersburg, FL  03/31/1998     NaN   
42                   T-Mobile Park         Seattle, WA  07/15/1999     NaN   
43                     Oracle Park   San Francisco, CA  04/11/2000     NaN   
44                   Comerica Park         Detroit, MI  04/11/2000     NaN   
45                Minute Maid Park         Houston, TX  04/07/2000     NaN   
46           American Family Field       Milwaukee, WI  04/06/2001     NaN   
47                        PNC Park      Pittsburgh, PA  04/09/2001     NaN   
52         Great American Ballpark      Cincinnati, OH  03/31/2003     NaN   
54              Citizens Bank Park    Philadelphia, PA  04/12/2004     NaN   
56                      Petco Park       San Diego, CA  04/08/2004     NaN   
60               Busch Stadium III       St. Louis, MO  04/10/2006     NaN   
62                  Nationals Park      Washington, DC  03/30/2008     NaN   
64                      Citi Field        New York, NY  04/13/2009     NaN   
65              New Yankee Stadium        New York, NY  04/16/2009     NaN   
67                    Target Field     Minneapolis, MN  04/12/2010     NaN   
70                  loanDepot Park           Miami, FL  04/04/2012     NaN   
90                     Truist Park         Atlanta, GA  04/14/2017     NaN   
104               Globe Life Field       Arlington, TX  07/24/2020     NaN   
212  Queen Elizabeth Olympic Pa...          London, EN  06/29/2019     NaN   
213            Muncy Bank Ballpark    Williamsport, PA  08/20/2017     NaN   
233      Estadio Alfredo Harp Helu    Mexico City, MEX  04/29/2023     NaN   
259             Ft. Street Grounds        St. Paul, MN         NaN     NaN   
260                           NAME         CITY, STATE         NaN     NaN   

     Seasons   Games    Lat    Long   Alt Current Historic  
0        112  8771.0  42.35  -71.10    16     Map      Map  
1        110  8634.0  41.95  -87.66   599     Map      Map  
7         62  5230.0  34.07 -118.24   501     Map      Map  
9         58  4595.0  33.80 -117.88   154     Map      Map  
12        56  4414.0  37.75 -122.20     0     Map      Map  
14        51  4009.0  39.05  -94.48   886     Map      Map  
26        34  2647.0  43.64  -79.39   268     Map      Map  
28        33  2587.0  41.83  -87.63   594     Map      Map  
30        32  2505.0  39.28  -76.62    35     Map      Map  
34        30  2333.0  41.50  -81.69   673     Map      Map  
35        29  2290.0  39.76 -104.99  5197     Map      Map  
39        26  2053.0  33.45 -112.07  1059     Map      Map  
40        26  2047.0  27.77  -82.65    41     Map      Map  
42        25  1934.0  47.59 -122.33    17     Map      Map  
43        24  1898.0  37.78 -122.39     8     Map      Map  
44        24  1889.0  42.34  -83.05   602     Map      Map  
45        24  1885.0  29.76  -95.36    20     Map      Map  
46        23  1819.0  43.03  -87.97   602     Map      Map  
47        23  1810.0  40.45  -80.01   724     Map      Map

Here we see that we have successfully removed all non-active ballparks from the steam DataFrame. There are more than 30 rows is this DataFrame because there are additional neutral site ballparks included in this DataFrame. 

During the course of this project steamheads.com stopped paying for their URL and the information was no longer able to be accessed. As a result I was able to pull the data from a previous version in my GitHub repository. Below we will re read in this steam data. 

In [151]:
# Import packages. 
import pandas as pd 

In [152]:
# Read in steam data from backup xlsx file. 
steam = pd.read_excel("steamheads_data_saved.xlsx")

# Display to verify the file read in properly. 
steam

Ballpark            Location   First Gm  Last Gm  \
0                     Fenway Park          Boston, MA 1912-04-20      NaN   
1                   Wrigley Field         Chicago, IL 1914-04-23      NaN   
2                  Dodger Stadium     Los Angeles, CA 1962-04-10      NaN   
3        Angel Stadium of Anaheim         Anaheim, CA 1966-04-19      NaN   
4   Oakland-Alameda County Col...         Oakland, CA 1968-04-17      NaN   
5                Kauffman Stadium     Kansas City, MO 1973-04-10      NaN   
6                   Rogers Centre        Toronto, ONT 1989-06-05      NaN   
7             Guarantee Rate Park         Chicago, IL 1991-04-18      NaN   
8   Oriole Park at Camden Yard...       Baltimore, MD 1992-04-06      NaN   
9               Progressive Field       Cleveland, OH 1994-04-04      NaN   
10                    Coors Field          Denver, CO 1995-04-26      NaN   
11                    Chase Field         Phoenix, AZ 1998-03-31      NaN   
12                Tropicana Field  St. Petersburg, FL 1998-03-31      NaN   
13                  T-Mobile Park         Seattle, WA 1999-07-15      NaN   
14                    Oracle Park   San Francisco, CA 2000-04-11      NaN   
15                  Comerica Park         Detroit, MI 2000-04-11      NaN   
16               Minute Maid Park         Houston, TX 2000-04-07      NaN   
17          American Family Field       Milwaukee, WI 2001-04-06      NaN   
18                       PNC Park      Pittsburgh, PA 2001-04-09      NaN   
19        Great American Ballpark      Cincinnati, OH 2003-03-31      NaN   
20             Citizens Bank Park    Philadelphia, PA 2004-04-12      NaN   
21                     Petco Park       San Diego, CA 2004-04-08      NaN   
22              Busch Stadium III       St. Louis, MO 2006-04-10      NaN   
23                 Nationals Park      Washington, DC 2008-03-30      NaN   
24                     Citi Field        New York, NY 2009-04-13      NaN   
25             New Yankee Stadium        New York, NY 2009-04-16      NaN   
26                   Target Field     Minneapolis, MN 2010-04-12      NaN   
27                 loanDepot Park           Miami, FL 2012-04-04      NaN   
28                    Truist Park         Atlanta, GA 2017-04-14      NaN   
29               Globe Life Field       Arlington, TX 2020-07-24      NaN   
30  Queen Elizabeth Olympic Pa...          London, EN 2019-06-29      NaN   
31            Muncy Bank Ballpark    Williamsport, PA 2017-08-20      NaN   
32      Estadio Alfredo Harp Helu    Mexico City, MEX 2023-04-29      NaN   
33             Ft. Street Grounds        St. Paul, MN        NaT      NaN   

    Seasons  Games    Lat    Long   Alt Current Historic  
0       112   8771  42.35  -71.10    16     Map      Map  
1       110   8634  41.95  -87.66   599     Map      Map  
2        62   5230  34.07 -118.24   501     Map      Map  
3        58   4595  33.80 -117.88   154     Map      Map  
4        56   4414  37.75 -122.20     0     Map      Map  
5        51   4009  39.05  -94.48   886     Map      Map  
6        34   2647  43.64  -79.39   268     Map      Map  
7        33   2587  41.83  -87.63   594     Map      Map  
8        32   2505  39.28  -76.62    35     Map      Map  
9        30   2333  41.50  -81.69   673     Map      Map  
10       29   2290  39.76 -104.99  5197     Map      Map  
11       26   2053  33.45 -112.07  1059     Map      Map  
12       26   2047  27.77  -82.65    41     Map      Map  
13       25   1934  47.59 -122.33    17     Map      Map  
14       24   1898  37.78 -122.39     8     Map      Map  
15       24   1889  42.34  -83.05   602     Map      Map  
16       24   1885  29.76  -95.36    20     Map      Map  
17       23   1819  43.03  -87.97   602     Map      Map  
18       23   1810  40.45  -80.01   724     Map      Map  
19       21   1646  39.10  -84.51   489     Map      Map  
20       20   1573  39.91  -75.17    19     Map      Map  
21       20   1566  32.71 -117.

# Transformation 3: Perform Fuzzy Matching on "Name" and "Ballpark" 

Performing the fuzzy matching between the two keys from the wiki and steam DataFrames will allow us to merge the DataFrames into one in a later step. To do this we need to create matching keys in each DataFrame. 

In [154]:
# Import Packages. 
import pandas as pd
from fuzzywuzzy import fuzz
from fuzzywuzzy import process

In [155]:
# Apply fuzzy matching to create matching name column in each df. 

# Use the process.extractOne method from fuzzywuzzy and token_sort_ratio to find the best match for each name in a given DataFrame. 
# use the apply method to assign the matched name to the new column 'matched_name'
wiki['matched_name'] = wiki['Name'].apply(lambda x: process.extractOne(x, steam['Ballpark'], scorer=fuzz.token_sort_ratio)[0])
steam['matched_name'] = steam['Ballpark'].apply(lambda x: process.extractOne(x, wiki['Name'], scorer=fuzz.token_sort_ratio)[0])

In [156]:
# Verify the results by printing a section of wiki. 

# Head wiki
wiki


Image                             Name    Capacity  \
0     NaN           American Family Field‡   41,900[1]   
1     NaN                    Angel Stadium   45,517[2]   
2     NaN                    Busch Stadium   45,494[3]   
3     NaN                     Chase Field‡   48,405[4]   
4     NaN                       Citi Field   41,922[5]   
5     NaN               Citizens Bank Park   42,792[6]   
6     NaN                    Comerica Park   41,083[7]   
7     NaN                      Coors Field   50,144[8]   
8     NaN                   Dodger Stadium   56,000[9]   
9     NaN                      Fenway Park  37,755[10]   
10    NaN                Globe Life Field‡  40,300[11]   
11    NaN         Great American Ball Park       43500   
12    NaN            Guaranteed Rate Field       40615   
13    NaN                 Kauffman Stadium  37,903[12]   
14    NaN                  LoanDepot Park‡       36742   
15    NaN                Minute Maid Park‡  41,168[14]   
16    NaN                   Nationals Park  41,373[16]   
17    NaN  Oakland–Alameda County Coliseum  46,847[17]   
18    NaN                      Oracle Park  41,265[18]   
19    NaN      Oriole Park at Camden Yards       44970   
20    NaN                       Petco Park  39,860[19]   
21    NaN                         PNC Park  38,747[20]   
22    NaN                Progressive Field  34,830[21]   
23    NaN                   Rogers Centre‡  39,150[22]   
24    NaN                   T-Mobile Park‡  47,929[23]   
25    NaN                     Target Field  38,544[24]   
26    NaN                 Tropicana Field†  25,000[25]   
27    NaN                      Truist Park  41,084[26]   
28    NaN                    Wrigley Field  41,649[27]   
29    NaN                   Yankee Stadium  46,537[28]   

                      Location          Surface                       Team  \
0         Milwaukee, Wisconsin            Grass          Milwaukee Brewers   
1          Anaheim, California            Grass         Los Angeles Angels   
2          St. Louis, Missouri            Grass        St. Louis Cardinals   
3             Phoenix, Arizona  Artificial turf       Arizona Diamondbacks   
4             Queens, New York            Grass              New York Mets   
5   Philadelphia, Pennsylvania            Grass      Philadelphia Phillies   
6            Detroit, Michigan            Grass             Detroit Tigers   
7             Denver, Colorado            Grass           Colorado Rockies   
8      Los Angeles, California            Grass  Los Angeles Dodgers[nb 2]   
9        Boston, Massachusetts            Grass       Boston Red Sox[nb 3]   
10            Arlington, Texas  Artificial turf              Texas Rangers   
11            Cincinnati, Ohio            Grass            Cincinnati Reds   
12           Chicago, Illinois            Grass          Chicago White Sox   
13       Kansas City, Missouri            Grass         Kansas City Royals   
14              Miami, Florida  Artificial turf              Miami Marlins   
15              Houston, Texas            Grass             Houston Astros   
16            Washington, D.C.            Grass       Washington Nationals   
17         Oakland, California            Grass          Oakland Athletics   
18   San Francisco, California            Grass       San Francisco Giants   
19         Baltimore, Maryland            Grass          Baltimore Orioles   
20       San Diego, California            Grass           San Diego Padres   
21    Pittsburgh, Pennsylvania            Grass         Pittsburgh Pirates   
22             Cleveland, Ohio            Grass        Cleveland Guardians   
23            Toronto, Ontario  Artificial turf          Toronto Blue Jays   
24         Seattle, Washington            Grass           Seattle Mariners   
25      Minneapolis, Minnesota            Grass            Minnesota Twins   
26     St. Petersburg, Florida  Artificial turf             Tampa Bay Rays   
27         Cumberland

Here we can see that everything matched well except the matched value for Angel Stadium. Lets Correct that by hand. 

In [158]:
# Use the Loc method to replace the matched value for Angel stadium. 
# Index 1
# Steam name = 'Angel Stadium of Anaheim'
wiki.loc[1, 'matched_name'] = 'Angel Stadium of Anaheim'

# verify the results . 
# Display first 5 rows. 
wiki.head(5)

Image                    Name   Capacity              Location  \
0    NaN  American Family Field‡  41,900[1]  Milwaukee, Wisconsin   
1    NaN           Angel Stadium  45,517[2]   Anaheim, California   
2    NaN           Busch Stadium  45,494[3]   St. Louis, Missouri   
3    NaN            Chase Field‡  48,405[4]      Phoenix, Arizona   
4    NaN              Citi Field  41,922[5]      Queens, New York   

           Surface                  Team Opened Distance to center field  \
0            Grass     Milwaukee Brewers   2001         400 feet (122 m)   
1            Grass    Los Angeles Angels   1966         396 feet (121 m)   
2            Grass   St. Louis Cardinals   2006         400 feet (122 m)   
3  Artificial turf  Arizona Diamondbacks   1998         407 feet (124 m)   
4            Grass         New York Mets   2009         408 feet (124 m)   

                  Type    Roof type              matched_name  
0         Retro-modern  Retractable     American Family Field  
1  Modern Retro-modern         Open  Angel Stadium of Anaheim  
2        Retro-classic         Open         Busch Stadium III  
3         Retro-modern  Retractable               Chase Field  
4        Retro-classic         Open                Citi Field

Now we have successfully completed a fuzzy matching. This work will allow us to move on to our next transformation and join the two DataFrames together. 

# Transformation 4: Merge both the wiki and steam DataFrames together. 

Here we will use a right outer join in order to add the information from the steam DataFrame to the wiki DataFrame. This join will be done on "Ballpark" and "matched_name" (from the wiki DataFrame) respectively. 

In [161]:
# Use merge method to join both datasets.
# Use right outer join.
# Keys: wiki = "matched_name", steam = "Ballpark". 
ballpark = wiki.merge(steam, how='inner', left_on='matched_name', right_on='Ballpark')

In [162]:
# Display the new dataset 
ballpark

Image                             Name    Capacity  \
0     NaN           American Family Field‡   41,900[1]   
1     NaN                    Angel Stadium   45,517[2]   
2     NaN                    Busch Stadium   45,494[3]   
3     NaN                     Chase Field‡   48,405[4]   
4     NaN                       Citi Field   41,922[5]   
5     NaN               Citizens Bank Park   42,792[6]   
6     NaN                    Comerica Park   41,083[7]   
7     NaN                      Coors Field   50,144[8]   
8     NaN                   Dodger Stadium   56,000[9]   
9     NaN                      Fenway Park  37,755[10]   
10    NaN                Globe Life Field‡  40,300[11]   
11    NaN         Great American Ball Park       43500   
12    NaN            Guaranteed Rate Field       40615   
13    NaN                 Kauffman Stadium  37,903[12]   
14    NaN                  LoanDepot Park‡       36742   
15    NaN                Minute Maid Park‡  41,168[14]   
16    NaN                   Nationals Park  41,373[16]   
17    NaN  Oakland–Alameda County Coliseum  46,847[17]   
18    NaN                      Oracle Park  41,265[18]   
19    NaN      Oriole Park at Camden Yards       44970   
20    NaN                       Petco Park  39,860[19]   
21    NaN                         PNC Park  38,747[20]   
22    NaN                Progressive Field  34,830[21]   
23    NaN                   Rogers Centre‡  39,150[22]   
24    NaN                   T-Mobile Park‡  47,929[23]   
25    NaN                     Target Field  38,544[24]   
26    NaN                 Tropicana Field†  25,000[25]   
27    NaN                      Truist Park  41,084[26]   
28    NaN                    Wrigley Field  41,649[27]   
29    NaN                   Yankee Stadium  46,537[28]   

                    Location_x          Surface                       Team  \
0         Milwaukee, Wisconsin            Grass          Milwaukee Brewers   
1          Anaheim, California            Grass         Los Angeles Angels   
2          St. Louis, Missouri            Grass        St. Louis Cardinals   
3             Phoenix, Arizona  Artificial turf       Arizona Diamondbacks   
4             Queens, New York            Grass              New York Mets   
5   Philadelphia, Pennsylvania            Grass      Philadelphia Phillies   
6            Detroit, Michigan            Grass             Detroit Tigers   
7             Denver, Colorado            Grass           Colorado Rockies   
8      Los Angeles, California            Grass  Los Angeles Dodgers[nb 2]   
9        Boston, Massachusetts            Grass       Boston Red Sox[nb 3]   
10            Arlington, Texas  Artificial turf              Texas Rangers   
11            Cincinnati, Ohio            Grass            Cincinnati Reds   
12           Chicago, Illinois            Grass          Chicago White Sox   
13       Kansas City, Missouri            Grass         Kansas City Royals   
14              Miami, Florida  Artificial turf              Miami Marlins   
15              Houston, Texas            Grass             Houston Astros   
16            Washington, D.C.            Grass       Washington Nationals   
17         Oakland, California            Grass          Oakland Athletics   
18   San Francisco, California            Grass       San Francisco Giants   
19         Baltimore, Maryland            Grass          Baltimore Orioles   
20       San Diego, California            Grass           San Diego Padres   
21    Pittsburgh, Pennsylvania            Grass         Pittsburgh Pirates   
22             Cleveland, Ohio            Grass        Cleveland Guardians   
23            Toronto, Ontario  Artificial turf          Toronto Blue Jays   
24         Seattle, Washington            Grass           Seattle Mariners   
25      Minneapolis, Minnesota            Grass            Minnesota Twins   
26     St. Petersburg, Florida  Artificial turf             Tampa Bay Rays   
27         Cumberland

We have successfully completed a join of the two separate DataFrames from our websites Wikipedia and Steamheads. Now working with one file we can begin to clean the DataFrame to make it more workable. Additionally we know that we have accounded for information from all 30  active Major League ballparks. 

# Transformation 5: Remove Unnecessary Columns

Removing unnecessary columns will allow us to remove both null columns, repetitive columns and columns that will not be necessary for the scope of our project. This will improve the readability and usability of the DataFrame. 

In [165]:
# First verify the size of the current data frame. 
# use shape 
print(ballpark.shape)

(30, 23)


In [166]:
# Use dropna to drop any columns that are completely empty. 
ballpark.dropna(axis=1, how='all', inplace=True)

# Verify the shape of ballpark. 
print(ballpark.shape)

(30, 21)


Here we see that we have sucessfully dropped 2 columns that contained only null data. 

Next we will print all column names and eliminate that columns that are not necessary. 

In [168]:
# Loop through all the columns in df ballpark. 
for column in ballpark.columns:
    # Print each column name
    print(column)

Name
Capacity
Location_x
Surface
Team
Opened
Distance to center field
Type
Roof type
matched_name_x
Ballpark
Location_y
First Gm
Seasons
Games
Lat
Long
Alt
Current
Historic
matched_name_y


Now we will drop the unnecessary columns. 

Columns to be dropped: Capacity, Opened, Distance to center field, Type, First Gm, Seasons, Games, Current, and Historic. 

In [170]:
# Create a list of columns to be removed. 
columns_to_remove = ['Capacity', 'Opened', 'Distance to center field', 'Type', 'First Gm', 'Seasons', 'Games', 'Current', 'Historic']

# Use drop to drop the columns in list. 
# Use Axis =1 
ballpark = ballpark.drop(columns=columns_to_remove, axis=1)

In [171]:
# Preview new DataFrame
ballpark.head(5)

Name            Location_x          Surface  \
0  American Family Field‡  Milwaukee, Wisconsin            Grass   
1           Angel Stadium   Anaheim, California            Grass   
2           Busch Stadium   St. Louis, Missouri            Grass   
3            Chase Field‡      Phoenix, Arizona  Artificial turf   
4              Citi Field      Queens, New York            Grass   

                   Team    Roof type            matched_name_x  \
0     Milwaukee Brewers  Retractable     American Family Field   
1    Los Angeles Angels         Open  Angel Stadium of Anaheim   
2   St. Louis Cardinals         Open         Busch Stadium III   
3  Arizona Diamondbacks  Retractable               Chase Field   
4         New York Mets         Open                Citi Field   

                   Ballpark     Location_y    Lat    Long   Alt  \
0     American Family Field  Milwaukee, WI  43.03  -87.97   602   
1  Angel Stadium of Anaheim    Anaheim, CA  33.80 -117.88   154   
2         Busch Stadium III  St. Louis, MO  38.62  -90.19   438   
3               Chase Field    Phoenix, AZ  33.45 -112.07  1059   
4                Citi Field   New York, NY  40.76  -73.85    13   

           matched_name_y  
0  American Family Field‡  
1           Angel Stadium  
2           Busch Stadium  
3            Chase Field‡  
4              Citi Field

In [172]:
# Remove multiple name and location columns. 

# Create a list of columns to be removed. 
columns_to_remove = ['matched_name_x','Location_y', 'matched_name_y', 'Ballpark' ]

# Use drop to drop the columns in list. 
# Use Axis =1 
ballpark = ballpark.drop(columns=columns_to_remove, axis=1)

In [173]:
# Preview new DataFrame
ballpark.head(5)

Name            Location_x          Surface  \
0  American Family Field‡  Milwaukee, Wisconsin            Grass   
1           Angel Stadium   Anaheim, California            Grass   
2           Busch Stadium   St. Louis, Missouri            Grass   
3            Chase Field‡      Phoenix, Arizona  Artificial turf   
4              Citi Field      Queens, New York            Grass   

                   Team    Roof type    Lat    Long   Alt  
0     Milwaukee Brewers  Retractable  43.03  -87.97   602  
1    Los Angeles Angels         Open  33.80 -117.88   154  
2   St. Louis Cardinals         Open  38.62  -90.19   438  
3  Arizona Diamondbacks  Retractable  33.45 -112.07  1059  
4         New York Mets         Open  40.76  -73.85    13

We Have successfully eliminated the unnecessary columns from the DataFrame. 

# Transformation 6: Rename Columns

Now we will rename the column names for several of the columns. This will improve readability and eliminate confusion at later steps. 

In [179]:
# Use Rename to rename necessary columns. 

# Change Name to Ballpark
ballpark.rename(columns={'Name': 'Ballpark'}, inplace=True)

# Change Location_x to City, 
ballpark.rename(columns={'Location_x': 'City'}, inplace=True)

In [181]:
## Preview new DataFrame
ballpark.head(5)

Ballpark                  City          Surface  \
0  American Family Field‡  Milwaukee, Wisconsin            Grass   
1           Angel Stadium   Anaheim, California            Grass   
2           Busch Stadium   St. Louis, Missouri            Grass   
3            Chase Field‡      Phoenix, Arizona  Artificial turf   
4              Citi Field      Queens, New York            Grass   

                   Team    Roof type    Lat    Long   Alt  
0     Milwaukee Brewers  Retractable  43.03  -87.97   602  
1    Los Angeles Angels         Open  33.80 -117.88   154  
2   St. Louis Cardinals         Open  38.62  -90.19   438  
3  Arizona Diamondbacks  Retractable  33.45 -112.07  1059  
4         New York Mets         Open  40.76  -73.85    13

# Transformation 7: Remove Dagger Mark from ballpark. 

Much like before we need to remove the dagger mark from some values in the Ballpark column. This will improve readability. 

In [186]:
# Use the str.replace method to remove the dagger character from every value in the "Ballpark" column. 
# Use ballpark dataset
ballpark['Ballpark'] = ballpark['Ballpark'].str.replace('‡', '')

We Have now completed all necessary transformations to our dataset. by removing the dagger mark. 

# Display Final Dataset

In [192]:
# Display the final dataset. 
ballpark

Ballpark                        City  \
0             American Family Field        Milwaukee, Wisconsin   
1                     Angel Stadium         Anaheim, California   
2                     Busch Stadium         St. Louis, Missouri   
3                       Chase Field            Phoenix, Arizona   
4                        Citi Field            Queens, New York   
5                Citizens Bank Park  Philadelphia, Pennsylvania   
6                     Comerica Park           Detroit, Michigan   
7                       Coors Field            Denver, Colorado   
8                    Dodger Stadium     Los Angeles, California   
9                       Fenway Park       Boston, Massachusetts   
10                 Globe Life Field            Arlington, Texas   
11         Great American Ball Park            Cincinnati, Ohio   
12            Guaranteed Rate Field           Chicago, Illinois   
13                 Kauffman Stadium       Kansas City, Missouri   
14                   LoanDepot Park              Miami, Florida   
15                 Minute Maid Park              Houston, Texas   
16                   Nationals Park            Washington, D.C.   
17  Oakland–Alameda County Coliseum         Oakland, California   
18                      Oracle Park   San Francisco, California   
19      Oriole Park at Camden Yards         Baltimore, Maryland   
20                       Petco Park       San Diego, California   
21                         PNC Park    Pittsburgh, Pennsylvania   
22                Progressive Field             Cleveland, Ohio   
23                    Rogers Centre            Toronto, Ontario   
24                    T-Mobile Park         Seattle, Washington   
25                     Target Field      Minneapolis, Minnesota   
26                 Tropicana Field†     St. Petersburg, Florida   
27                      Truist Park         Cumberland, Georgia   
28                    Wrigley Field           Chicago, Illinois   
29                   Yankee Stadium             Bronx, New York   

            Surface                       Team    Roof type    Lat    Long  \
0             Grass          Milwaukee Brewers  Retractable  43.03  -87.97   
1             Grass         Los Angeles Angels         Open  33.80 -117.88   
2             Grass        St. Louis Cardinals         Open  38.62  -90.19   
3   Artificial turf       Arizona Diamondbacks  Retractable  33.45 -112.07   
4             Grass              New York Mets         Open  40.76  -73.85   
5             Grass      Philadelphia Phillies         Open  39.91  -75.17   
6             Grass             Detroit Tigers         Open  42.34  -83.05   
7             Grass           Colorado Rockies         Open  39.76 -104.99   
8             Grass  Los Angeles Dodgers[nb 2]         Open  34.07 -118.24   
9             Grass       Boston Red Sox[nb 3]         Open  42.35  -71.10   
10  Artificial turf              Texas Rangers  Retractable  32.75  -97.08   
11            Grass            Cincinnati Reds         Open  39.10  -84.51   
12            Grass          Chicago White Sox         Open  41.83  -87.63   
13            Grass         Kansas City Royals         Open  39.05  -94.48   
14  Artificial turf              Miami Marlins  Retractable  25.78  -80.22   
15            Grass             Houston Astros  Retractable  29.76  -95.36   
16            Grass       Washington Nationals         Open  38.87  -77.01   
17            Grass          Oakland Athletics         Open  37.75 -122.20   
18            Grass       San Francisco Giants         Open  37.78 -122.39   
19            Grass          Baltimore Orioles         Open  39.28  -76.62   
20            Grass           San Diego Padres         Open  32.71 -117.16   
21            Grass         Pittsburgh Pirates         Open  40.45  -80.01   
22            Grass        Cleveland Guardians         Open  41.50  -81.69   
23  Artificial turf          Toronto Blue Jays  Retractable  43.64  -79.39   